### ISSUE TITLE
install_tensorflow (no envname) installs into loaded env (not r-tensorflow)

### ISSUE THREAD

**Issue by skeydan** [rstudio/tensorflow#316]
It seems like there was a change in install behavior that can make a thing like this happen (just quickly putting this here for further investigation):

- we have `condaenv1` loaded
- we call `install_tensorflow` with no specification of env, assuming this will install into `r-tensorflow`
- TF gets installed into loaded env instead

**Comment by skeydan** [comment]
Hi @kevinushey, could I involve you in this?
This change in behavior ought to be due to changes in `reticulate` because the `install_tensorflow()` procedure was not changed...
I can probably reproduce the above and provide logs, but first, a more detailed description of what probably happened:

As a preamble, I often use `install_tensorflow(version = "x")` as a lazy way to test with different TF versions.
With no other arguments than "version", this will erase the `r-tensorflow` env and create it anew, with the desired TF.

Yesterday what happened was (with a fresh `reticulate` from master):

1. I try to install the current nightly, which is broken. Now I don't have a functioning `r-tensorflow`.
2. For some reason (don't remember exactly), I end up using another env in rstudio, say `env2`. 
3. Now from `env2` I again try `install_tensorflow(version = "some previous nightly") `. 
   The installation process tells me `installing into r-tensorflow` as expected but then I see messages like

`requirement xy already satisfied in <path/to/env2>`

which currently is loaded but where we don't want to install...
And finally I see that really, the new binaries were installed in `env2` not `r-tensorflow`.

Bottom line, the user of the function expects the install to go to `r-tensorflow` (or at least one should expect that from the function interface). Now evidently what counts is the loaded env, right?

The question then is, if this is the intended behavior in `reticulate` now, how do we adapt `tensorflow` resp. how do we make them play together?

**Comment by kevinushey** [comment]
What's the simplest way for me to reproduce what you're seeing?

**Comment by skeydan** [comment]
This should work (attaching relevant parts of output).

```
######  install different env ######
reticulate::use_condaenv("r-tensorflow")
tensorflow::install_tensorflow(envname = "tmp")


###### activate that env, and do default install ######
reticulate::use_condaenv("tmp")
# not necessary, but makes it easier to see that tf gets installed into tmp
reticulate::conda_remove("r-tensorflow")
# default is "r-tensorflow"
tensorflow::install_tensorflow() 


###### checks ######
#  r-tensorflow exists
reticulate::conda_list() 
#  but: 
# conda activate r-tensorflow && conda list
# shows t-tensorflow has no tf packages
```

Output:

```
Creating r-tensorflow conda environment for TensorFlow installation...
Collecting package metadata: ...working... done
Solving environment: ...working... done

## Package Plan ##

  environment location: /home/key/anaconda3/envs/r-tensorflow
  ```
  
  ...
  ...
  
  ```
  Requirement already up-to-date: keras in /home/key/anaconda3/envs/tmp/lib/python3.7/site-packages (2.2.4)
Requirement already up-to-date: tensorflow-hub in /home/key/anaconda3/envs/tmp/lib/python3.7/site-packages (0.4.0)
  ```

Thanks!

**Comment by kevinushey** [comment]
Thanks! I'm OOF most of today but will take a look at this tomorrow.

**Comment by kevinushey** [comment]
It seems to work as expected for me:

```
tensorflow::install_tensorflow(envname = "tmp", method = "conda")
reticulate::use_condaenv("tmp", required = TRUE)
reticulate::conda_remove("r-tensorflow")
tensorflow::install_tensorflow(method = "conda")
system("/home/kevin/anaconda3/bin/conda list -n r-tensorflow")
```

I see:

```
Requirement already satisfied, skipping upgrade: six>=1.9.0 in /home/kevin/anaconda3/envs/r-tensorflow/lib/python3.7/site-packages (from keras) (1.12.0)
Requirement already satisfied, skipping upgrade: pyyaml in /home/kevin/anaconda3/envs/r-tensorflow/lib/python3.7/site-packages (from keras) (5.1)
Requirement already satisfied, skipping upgrade: h5py in /home/kevin/anaconda3/envs/r-tensorflow/lib/python3.7/site-packages (from keras) (2.9.0)
Requirement already satisfied, skipping upgrade: keras-preprocessing>=1.0.5 in /home/kevin/anaconda3/envs/r-tensorflow/lib/python3.7/site-packages (from keras) (1.0.9)
Requirement already satisfied, skipping upgrade: keras-applications>=1.0.6 in /home/kevin/anaconda3/envs/r-tensorflow/lib/python3.7/site-packages (from keras) (1.0.7)
Requirement already satisfied, skipping upgrade: scipy>=0.14 in /home/kevin/anaconda3/envs/r-tensorflow/lib/python3.7/site-packages (from keras) (1.2.1)
Requirement already satisfied, skipping upgrade: numpy>=1.9.1 in /home/kevin/anaconda3/envs/r-tensorflow/lib/python3.7/site-packages (from keras) (1.16.2)
Requirement already satisfied, skipping upgrade: protobuf>=3.4.0 in /home/kevin/anaconda3/envs/r-tensorflow/lib/python3.7/site-packages (from tensorflow-hub) (3.7.1)
Requirement already satisfied, skipping upgrade: setuptools in /home/kevin/anaconda3/envs/r-tensorflow/lib/python3.7/site-packages (from protobuf>=3.4.0->tensorflow-hub) (41.0.0)
Installing collected packages: keras, tensorflow-hub, tensorflow-probability
Successfully installed keras-2.2.4 tensorflow-hub-0.4.0 tensorflow-probability-0.6.0
```

It's also not clear to me why your call to e.g. `install_tensorflow()` without an explicit `method` argument uses conda environment rather than a virtual environment.

**Comment by kevinushey** [comment]
Are you doing anything like setting a default version of Python (`RETICULATE_PYTHON`) in a startup R profile or `.Renviron` or anything like that?

**Comment by skeydan** [comment]
Thanks Kevin!
No, nothing in R startup files. I have this my `.bashrc` though

```
# >>> conda initialize >>>
# !! Contents within this block are managed by 'conda init' !!
__conda_setup="$('/home/key/anaconda3/bin/conda' 'shell.bash' 'hook' 2> /dev/null)"
if [ $? -eq 0 ]; then
    eval "$__conda_setup"
else
    if [ -f "/home/key/anaconda3/etc/profile.d/conda.sh" ]; then
        . "/home/key/anaconda3/etc/profile.d/conda.sh"
    else
        export PATH="/home/key/anaconda3/bin:$PATH"
    fi
fi
unset __conda_setup
```

and thus, this on the PATH

```
~/anaconda3/condabin
```

and this alias

```
CONDA_PYTHON_EXE=/~/anaconda3/bin/python
```

I am really surprised repro doesn't work for you...

**Comment by skeydan** [comment]
btw my conda version is `4.6.7`, and I'm pretty sure that update changed the entry in `.bashrc` (and for me after a long period of not updating `conda`, this introduced the `condabin` thing...)

**Comment by kevinushey** [comment]
Still no luck reproducing, both with the CRAN and development versions of `reticulate` / `tensorflow`.

The only relevant change I can think of in `reticulate` is related to `RETICULATE_PYTHON_ENV` -- if that environment variable is set, then it's used as the default environment rather than `r-tensorflow`. However, `reticulate` never actually sets that (the only place where we play with that right now is `renv`).

**Comment by skeydan** [comment]
Hi @kevinushey thanks for retrying. I just did, too, using your exact commands 

```
tensorflow::install_tensorflow(envname = "tmp", method = "conda")
reticulate::use_condaenv("tmp", required = TRUE)
reticulate::conda_remove("r-tensorflow")
tensorflow::install_tensorflow(method = "conda")
system("/home/key/anaconda3/bin/conda list -n r-tensorflow")
```

and after updating `reticulate` and `tensorflow` from github, and `conda` to v 4.6.14.

And I still see the same behavior: no TF installed in `r-tensorflow`.

When I query the PATH from inside rstudio after those commands, I see that `/home/key/anaconda3/envs/tmp/bin` got prepended. I wonder if that could be related?

**Comment by skeydan** [comment]
Now I guess I'm really confused ... so I wanted to debug inside `install_tensorflow_conda`, and so I pulled from gh and built the package and re-executed those steps above ... and can't reproduce any more... (even though before I'd updated from github and so, `tensorflow` versions should have been the same..?)

I don't see that `tmp` env prepended to the PATH anymore either, so probably that was related.
I just have no idea why it did get prepended ...

**Comment by skeydan** [comment]
closing for non-reproducibility for now...

**Comment by skeydan** [comment]
Just happened again. Note i'm calling `install_tensorflow` with some `conda` env being loaded (implicitly assuming the R session will be restarted, or otherwise take care it doesn't make use of the env currently being loaded, - haven't checked yet if it is):

Here the ERROR lines at the end indicate it's installing into the wrong, previously-loaded, env, which has `tensorflow 2.0.0b1`:

```
> tensorflow::install_tensorflow(method = "conda", envname = "n", version = "nightly")
Creating  n  conda environment... 
Collecting package metadata: ...working... done
Solving environment: ...working... done

## Package Plan ##

  environment location: /home/key/anaconda3/envs/n

  added / updated specs:
    - python=3.6


The following NEW packages will be INSTALLED:

  _libgcc_mutex      pkgs/main/linux-64::_libgcc_mutex-0.1-main
  ca-certificates    pkgs/main/linux-64::ca-certificates-2019.5.15-0
  certifi            pkgs/main/linux-64::certifi-2019.6.16-py36_0
  libedit            pkgs/main/linux-64::libedit-3.1.20181209-hc058e9b_0
  libffi             pkgs/main/linux-64::libffi-3.2.1-hd88cf55_4
  libgcc-ng          pkgs/main/linux-64::libgcc-ng-9.1.0-hdf63c60_0
  libstdcxx-ng       pkgs/main/linux-64::libstdcxx-ng-9.1.0-hdf63c60_0
  ncurses            pkgs/main/linux-64::ncurses-6.1-he6710b0_1
  openssl            pkgs/main/linux-64::openssl-1.1.1c-h7b6447c_1
  pip                pkgs/main/linux-64::pip-19.1.1-py36_0
  python             pkgs/main/linux-64::python-3.6.8-h0371630_0
  readline           pkgs/main/linux-64::readline-7.0-h7b6447c_5
  setuptools         pkgs/main/linux-64::setuptools-41.0.1-py36_0
  sqlite             pkgs/main/linux-64::sqlite-3.28.0-h7b6447c_0
  tk                 pkgs/main/linux-64::tk-8.6.8-hbc83047_0
  wheel              pkgs/main/linux-64::wheel-0.33.4-py36_0
  xz                 pkgs/main/linux-64::xz-5.2.4-h14c3975_4
  zlib               pkgs/main/linux-64::zlib-1.2.11-h7b6447c_3


Preparing transaction: ...working... done
Verifying transaction: ...working... done
Executing transaction: ...working... done
#
# To activate this environment, use
#
#     $ conda activate n
#
# To deactivate an active environment, use
#
#     $ conda deactivate



==> WARNING: A newer version of conda exists. <==
  current version: 4.6.14
  latest version: 4.7.5

Please update conda by running

    $ conda update -n base -c defaults conda


Installing python modules...
Collecting tf-nightly
  Using cached https://files.pythonhosted.org/packages/fe/6f/9917f6777af70014574a07a8beb3b87b7353e2cdc9abb5472286ff32435c/tf_nightly-1.15.0.dev20190628-cp36-cp36m-manylinux1_x86_64.whl
Collecting tensorflow-hub
  Using cached https://files.pythonhosted.org/packages/b5/be/f18c352d84382d9c795a0f37eaf16d42ace7d161fbb0ad20bdcd5e550015/tensorflow_hub-0.5.0-py2.py3-none-any.whl
Collecting tfp-nightly
  Using cached https://files.pythonhosted.org/packages/68/e1/d7030c1a3e659a4a42c1c5b49902302ed5bcc00fd948aa3947713e9e03de/tfp_nightly-0.8.0.dev20190702-py2.py3-none-any.whl
Collecting gast>=0.2.0 (from tf-nightly)
Collecting tf-estimator-nightly (from tf-nightly)
  Using cached https://files.pythonhosted.org/packages/10/07/527b55522ad7c3b8b22ee00ba08d23057de506c465c94c8ac48dc7a1155b/tf_estimator_nightly-1.14.0.dev2019070201-py2.py3-none-any.whl
Collecting numpy<2.0,>=1.14.5 (from tf-nightly)
  Using cached https://files.pythonhosted.org/packages/87/2d/e4656149cbadd3a8a0369fcd1a9c7d61cc7b87b3903b85389c70c989a696/numpy-1.16.4-cp36-cp36m-manylinux1_x86_64.whl
Collecting wheel>=0.26 (from tf-nightly)
  Using cached https://files.pythonhosted.org/packages/bb/10/44230dd6bf3563b8f227dbf344c908d412ad2ff48066476672f3a72e174e/wheel-0.33.4-py2.py3-none-any.whl
Collecting keras-preprocessing>=1.0.5 (from tf-nightly)
  Using cached https://files.pythonhosted.org/packages/28/6a/8c1f62c37212d9fc441a7e26736df51ce6f0e38455816445471f10da4f0a/Keras_Preprocessing-1.1.0-py2.py3-none-any.whl
Collecting protobuf>=3.6.1 (from tf-nightly)
  Using cached https://files.pythonhosted.org/packages/d2/fb/29de8d08967f0cce1bb10b39846d836b0f3bf6776ddc36aed7c73498ca7e/protobuf-3.8.0-cp36-cp36m-manylinux1_x86_64.whl
Collecting keras-applications>=1.0.8 (from tf-nightly)
  Using cached https://files.pythonhosted.org/packages/71/e3/19762fdfc62877ae9102edf6342d71b28fbfd9dea3d2f96a882ce099b03f/Keras_Applications-1.0.8-py3-none-any.whl
Collecting astor>=0.6.0 (from tf-nightly)
  Using cached https://files.pythonhosted.org/packages/d1/4f/950dfae467b384fc96bc6469de25d832534f6b4441033c39f914efd13418/astor-0.8.0-py2.py3-none-any.whl
Collecting grpcio>=1.8.6 (from tf-nightly)
  Using cached https://files.pythonhosted.org/packages/f2/5d/b434403adb2db8853a97828d3d19f2032e79d630e0d11a8e95d243103a11/grpcio-1.22.0-cp36-cp36m-manylinux1_x86_64.whl
Collecting wrapt>=1.11.1 (from tf-nightly)
Collecting google-pasta>=0.1.6 (from tf-nightly)
  Using cached https://files.pythonhosted.org/packages/d0/33/376510eb8d6246f3c30545f416b2263eee461e40940c2a4413c711bdf62d/google_pasta-0.1.7-py3-none-any.whl
Collecting opt-einsum>=2.3.2 (from tf-nightly)
Collecting termcolor>=1.1.0 (from tf-nightly)
Collecting tb-nightly<1.15.0a0,>=1.14.0a0 (from tf-nightly)
  Using cached https://files.pythonhosted.org/packages/3a/60/afa129c3621d62c885599076f8e89737d66e5dfffad1a08842b1c11b4540/tb_nightly-1.14.0a20190614-py3-none-any.whl
Collecting absl-py>=0.7.0 (from tf-nightly)
Collecting six>=1.10.0 (from tf-nightly)
  Using cached https://files.pythonhosted.org/packages/73/fb/00a976f728d0d1fecfe898238ce23f502a721c0ac0ecfedb80e0d88c64e9/six-1.12.0-py2.py3-none-any.whl
Collecting decorator (from tfp-nightly)
  Using cached https://files.pythonhosted.org/packages/5f/88/0075e461560a1e750a0dcbf77f1d9de775028c37a19a346a6c565a257399/decorator-4.4.0-py2.py3-none-any.whl
Collecting cloudpickle==1.1.1 (from tfp-nightly)
  Using cached https://files.pythonhosted.org/packages/24/fb/4f92f8c0f40a0d728b4f3d5ec5ff84353e705d8ff5e3e447620ea98b06bd/cloudpickle-1.1.1-py2.py3-none-any.whl
Collecting setuptools (from protobuf>=3.6.1->tf-nightly)
  Using cached https://files.pythonhosted.org/packages/ec/51/f45cea425fd5cb0b0380f5b0f048ebc1da5b417e48d304838c02d6288a1e/setuptools-41.0.1-py2.py3-none-any.whl
Collecting h5py (from keras-applications>=1.0.8->tf-nightly)
  Using cached https://files.pythonhosted.org/packages/30/99/d7d4fbf2d02bb30fb76179911a250074b55b852d34e98dd452a9f394ac06/h5py-2.9.0-cp36-cp36m-manylinux1_x86_64.whl
Collecting werkzeug>=0.11.15 (from tb-nightly<1.15.0a0,>=1.14.0a0->tf-nightly)
  Using cached https://files.pythonhosted.org/packages/9f/57/92a497e38161ce40606c27a86759c6b92dd34fcdb33f64171ec559257c02/Werkzeug-0.15.4-py2.py3-none-any.whl
Collecting markdown>=2.6.8 (from tb-nightly<1.15.0a0,>=1.14.0a0->tf-nightly)
  Using cached https://files.pythonhosted.org/packages/c0/4e/fd492e91abdc2d2fcb70ef453064d980688762079397f779758e055f6575/Markdown-3.1.1-py2.py3-none-any.whl
ERROR: tensorflow 2.0.0b1 has requirement tb-nightly<1.14.0a20190604,>=1.14.0a20190603, but you'll have tb-nightly 1.14.0a20190614 which is incompatible.
ERROR: tensorflow 2.0.0b1 has requirement tf-estimator-nightly<1.14.0.dev2019060502,>=1.14.0.dev2019060501, but you'll have tf-estimator-nightly 1.14.0.dev2019070201 which is incompatible.
Installing collected packages: gast, tf-estimator-nightly, numpy, wheel, six, keras-preprocessing, setuptools, protobuf, h5py, keras-applications, astor, grpcio, wrapt, google-pasta, opt-einsum, termcolor, absl-py, werkzeug, markdown, tb-nightly, tf-nightly, tensorflow-hub, decorator, cloudpickle, tfp-nightly
Successfully installed absl-py-0.7.1 astor-0.8.0 cloudpickle-1.1.1 decorator-4.4.0 gast-0.2.2 google-pasta-0.1.7 grpcio-1.22.0 h5py-2.9.0 keras-applications-1.0.8 keras-preprocessing-1.1.0 markdown-3.1.1 numpy-1.16.4 opt-einsum-2.3.2 protobuf-3.8.0 setuptools-41.0.1 six-1.12.0 tb-nightly-1.14.0a20190614 tensorflow-hub-0.5.0 termcolor-1.1.0 tf-estimator-nightly-1.14.0.dev2019070201 tf-nightly-1.15.0.dev20190628 tfp-nightly-0.8.0.dev20190702 werkzeug-0.15.4 wheel-0.33.4 wrapt-1.11.2

Installation complete.


Restarting R session...

> 
```

**Comment by skeydan** [comment]
Hm, another confirming case.

I was working with a virtualenv loaded this time, and without session restart, tried to install a conda env:

```
> tensorflow::install_tensorflow(method = "conda", envname = "b", version = "2.0.0-beta1",conda_python_version = "2.7")
Removing  b  conda environment... 

## Package Plan ##

  environment location: /home/key/anaconda3/envs/b


The following packages will be REMOVED:

  _libgcc_mutex-0.1-main
  ca-certificates-2019.5.15-0
  certifi-2019.6.16-py36_0
  libedit-3.1.20181209-hc058e9b_0
  libffi-3.2.1-hd88cf55_4
  libgcc-ng-9.1.0-hdf63c60_0
  libstdcxx-ng-9.1.0-hdf63c60_0
  ncurses-6.1-he6710b0_1
  openssl-1.1.1c-h7b6447c_1
  pip-19.1.1-py36_0
  python-3.6.8-h0371630_0
  readline-7.0-h7b6447c_5
  setuptools-41.0.1-py36_0
  sqlite-3.28.0-h7b6447c_0
  tk-8.6.8-hbc83047_0
  wheel-0.33.4-py36_0
  xz-5.2.4-h14c3975_4
  zlib-1.2.11-h7b6447c_3


Preparing transaction: ...working... done
Verifying transaction: ...working... done
Executing transaction: ...working... done

Remove all packages in environment /home/key/anaconda3/envs/b:

Creating  b  conda environment... 
Collecting package metadata: ...working... done
Solving environment: ...working... done

## Package Plan ##

  environment location: /home/key/anaconda3/envs/b

  added / updated specs:
    - python=2.7


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    certifi-2019.6.16          |           py27_0         154 KB
    pip-19.1.1                 |           py27_0         1.8 MB
    python-2.7.16              |       h9bab390_0        12.8 MB
    setuptools-41.0.1          |           py27_0         640 KB
    wheel-0.33.4               |           py27_0          39 KB
    ------------------------------------------------------------
                                           Total:        15.4 MB

The following NEW packages will be INSTALLED:

  _libgcc_mutex      pkgs/main/linux-64::_libgcc_mutex-0.1-main
  ca-certificates    pkgs/main/linux-64::ca-certificates-2019.5.15-0
  certifi            pkgs/main/linux-64::certifi-2019.6.16-py27_0
  libedit            pkgs/main/linux-64::libedit-3.1.20181209-hc058e9b_0
  libffi             pkgs/main/linux-64::libffi-3.2.1-hd88cf55_4
  libgcc-ng          pkgs/main/linux-64::libgcc-ng-9.1.0-hdf63c60_0
  libstdcxx-ng       pkgs/main/linux-64::libstdcxx-ng-9.1.0-hdf63c60_0
  ncurses            pkgs/main/linux-64::ncurses-6.1-he6710b0_1
  openssl            pkgs/main/linux-64::openssl-1.1.1c-h7b6447c_1
  pip                pkgs/main/linux-64::pip-19.1.1-py27_0
  python             pkgs/main/linux-64::python-2.7.16-h9bab390_0
  readline           pkgs/main/linux-64::readline-7.0-h7b6447c_5
  setuptools         pkgs/main/linux-64::setuptools-41.0.1-py27_0
  sqlite             pkgs/main/linux-64::sqlite-3.28.0-h7b6447c_0
  tk                 pkgs/main/linux-64::tk-8.6.8-hbc83047_0
  wheel              pkgs/main/linux-64::wheel-0.33.4-py27_0
  zlib               pkgs/main/linux-64::zlib-1.2.11-h7b6447c_3



Downloading and Extracting Packages
certifi-2019.6.16    | 154 KB    | ########## | 100% 
python-2.7.16        | 12.8 MB   | ########## | 100% 
setuptools-41.0.1    | 640 KB    | ########## | 100% 
wheel-0.33.4         | 39 KB     | ########## | 100% 
pip-19.1.1           | 1.8 MB    | ########## | 100% 
Preparing transaction: ...working... done
Verifying transaction: ...working... done
Executing transaction: ...working... done
#
# To activate this environment, use
#
#     $ conda activate b
#
# To deactivate an active environment, use
#
#     $ conda deactivate



==> WARNING: A newer version of conda exists. <==
  current version: 4.6.14
  latest version: 4.7.5

Please update conda by running

    $ conda update -n base -c defaults conda


Installing python modules...
/bin/bash: /home/key/.virtualenvs/vb/bin/pip: /home/key/.virtualenvs/vb/bin/python3.6: bad interpreter: No such file or directory
 Error: Error 126 occurred installing packages into conda environment b 

```

**Comment by kevinushey** [comment]
Do we have a minimally reproducible example for this yet?

I wonder if the fact that we modify the PATH when loading a virtual environment is coming into play? (thereby forcing the version of pip etc. from that virtual environment to be used)

**Comment by skeydan** [comment]
Sorry for the delay, I was busy with a post and am at UseR! now -- I'll test with the revamped `py_install` soon...

**Comment by skeydan** [comment]
Given I already once unsuccessfully tried to reproduce (until weeks later, it "happened" again), and we're in the process of adapting `install_tensorflow` to more directly make use of `py_install` ... I think I'll close for now (again) and possibly reopen if it still happens after `install_tensorflow` has been fully refactored.
